File: src/core/engine/verify-shape.ts
/**
 * Check 1 of `compiled-program.md` 3.5: every required field is present and of
 * the declared type.
 *
 * This is the half of verification that reads an untrusted object and decides
 * whether it is a compiled program at all. It runs before anything else,
 * because every later check indexes into a table this one proves is a table.
 *
 * The failures are all one code. OS6018 covers a malformed instruction list, an
 * unreadable encoding and a field of the wrong type because they are not
 * separate fixes: each is a defect of the compiler that wrote the program, and
 * none of them is repairable by hand. What varies is the location, and the
 * location is a field path, so a compiler author is told `outputs.plots[2].
 * channel` rather than "structure".
 */
import type { Diagnostic } from '../diagnostics/index.js';
import { malformed } from './errors.js';

type Unknown = Readonly<Record<string, unknown>>;

/** Collects the first failure and stops: a malformed program has no second opinion. */
export class ShapeCheck {
  private failure: Diagnostic | undefined;

  problem(): Diagnostic | undefined {
    return this.failure;
  }

  fail(path: string, reason: string): false {
    this.failure ??= malformed(path, reason);
    return false;
  }

  object(value: unknown, path: string): value is Unknown {
    if (typeof value !== 'object' || value === null || Array.isArray(value)) {
      return this.fail(path, 'an object was required');
    }
    return true;
  }

  array(value: unknown, path: string): value is readonly unknown[] {
    if (!Array.isArray(value)) return this.fail(path, 'an array was required');
    return true;
  }

  number(value: unknown, path: string): value is number {
    if (typeof value !== 'number' || !Number.isFinite(value)) {
      return this.fail(path, 'a finite number was required');
    }
    return true;
  }

  whole(value: unknown, path: string): value is number {
    if (!this.number(value, path)) return false;
    if (!Number.isInteger(value)) return this.fail(path, 'a whole number was required');
    return true;
  }

  string(value: unknown, path: string): value is string {
    if (typeof value !== 'string') return this.fail(path, 'a string was required');
    return true;
  }

  bool(value: unknown, path: string): value is boolean {
    if (typeof value !== 'boolean') return this.fail(path, 'a true or false was required');
    return true;
  }

  /** A whole number that indexes a table, with the table's size in the message. */
  index(value: unknown, path: string, size: number, table: string): value is number {
    if (!this.whole(value, path)) return false;
    if (value < 0 || value >= size) {
      return this.fail(path, `${value} is outside ${table}, which holds ${size}`);
    }
    return true;
  }

  one<T extends string>(value: unknown, path: string, allowed: readonly T[]): value is T {
    if (!this.string(value, path)) return false;
    if (!(allowed as readonly string[]).includes(value)) {
      return this.fail(path, `${value} is not one of ${allowed.join(', ')}`);
    }
    return true;
  }

  /** A field that may be a value or null, which is a value the script could write. */
  nullable(value: unknown, path: string, check: (v: unknown, p: string) => boolean): boolean {
    return value === null ? true : check.call(this, value, path);
  }
}

/** The tags of a constant pool entry, 2.9. */
const CONSTANT_TAGS = ['z', 'b', 'n', 's', 'c'] as const;

export function checkConstant(shape: ShapeCheck, value: unknown, path: string): boolean {
  if (!shape.array(value, path)) return false;
  if (value.length !== 2) return shape.fail(path, 'a pool entry is a tag and a value');
  const tag = value[0];
  if (!shape.one(tag, `${path}[0]`, CONSTANT_TAGS)) return false;
  const held = value[1];
  switch (tag) {
    case 'z':
      return held === null ? true : shape.fail(`${path}[1]`, 'the absent entry holds null');
    case 'b':
      return shape.bool(held, `${path}[1]`);
    case 'n':
      return shape.number(held, `${path}[1]`);
    case 's':
      return shape.string(held, `${path}[1]`);
    default:
      return checkColour(shape, held, `${path}[1]`);
  }
}

export function checkColour(shape: ShapeCheck, value: unknown, path: string): boolean {
  if (!shape.array(value, path)) return false;
  if (value.length !== 4) return shape.fail(path, 'a colour is four numbers');
  for (let i = 0; i < 3; i += 1) {
    if (!shape.whole(value[i], `${path}[${i}]`)) return false;
    const channel = value[i] as number;
    if (channel < 0 || channel > 255) {
      return shape.fail(`${path}[${i}]`, `a channel runs 0 to 255 and this is ${channel}`);
    }
  }
  if (!shape.number(value[3], `${path}[3]`)) return false;
  const alpha = value[3] as number;
  if (alpha < 0 || alpha > 1) {
    return shape.fail(`${path}[3]`, `an alpha runs 0 to 1 and this is ${alpha}`);
  }
  return true;
}

/**
 * A declaration field, 2.3: a value, or the `{ "input": key }` reference.
 *
 * The key half of check 10 lives here, because this is the one walk that
 * visits every field that may hold one. The value half runs later, with the
 * host's settings in hand.
 */
export function checkField(
  shape: ShapeCheck,
  value: unknown,
  path: string,
  keys: ReadonlySet<string>,
): boolean {
  if (value === null) return true;
  const kind = typeof value;
  if (kind === 'boolean' || kind === 'string') return true;
  if (kind === 'number') return shape.number(value, path);
  if (Array.isArray(value)) {
    for (let i = 0; i < value.length; i += 1) {
      if (!shape.number(value[i], `${path}[${i}]`)) return false;
    }
    return true;
  }
  if (!shape.object(value, path)) return false;
  const key = value['input'];
  if (key === undefined) return shape.fail(path, 'an object here is an input reference');
  if (!shape.string(key, `${path}.input`)) return false;
  if (!keys.has(key)) {
    return shape.fail(path, `it names the input ${key}, which inputs[] does not declare`);
  }
  return true;
}

File: engine/openscript/verify_shape.py
"""Check 1 of ``compiled-program.md`` section 3.5: is this a compiled program at all.

This is the half of verification that reads an untrusted object and answers that
question before anything else runs, because every later check indexes into a
table this one proves is a table.

Every failure here is OS6018. A malformed instruction list, an unreadable
encoding and a field of the wrong type are not three fixes: each is a defect of
the compiler that wrote the program and none of them is repairable by hand. What
varies is the location, and the location is a field path, so whoever wrote the
compiler is told ``outputs.plots[2].channel`` rather than "structure".

One difference from the first engine, and it is this interpreter's alone: its
own ``True`` is an instance of its integer type, so every numeric test below
refuses a boolean explicitly. Without that line a program carrying ``true``
where a channel index belongs would verify, and the failure would arrive later
as a wrong number.
"""

from typing import Any, Callable, Optional, Sequence, Set

from .diagnostics import Diagnostic, malformed

#: The tags a constant pool entry may carry, section 2.9.
CONSTANT_TAGS = ("z", "b", "n", "s", "c")


class ShapeCheck:
    """Collects the first failure and stops: a malformed program has no second opinion."""

    def __init__(self) -> None:
        self._failure: Optional[Diagnostic] = None

    def problem(self) -> Optional[Diagnostic]:
        return self._failure

    def fail(self, path: str, reason: str) -> bool:
        if self._failure is None:
            self._failure = malformed(path, reason)
        return False

    def object(self, value: Any, path: str) -> bool:
        if not isinstance(value, dict):
            return self.fail(path, "an object was required")
        return True

    def array(self, value: Any, path: str) -> bool:
        if not isinstance(value, list):
            return self.fail(path, "an array was required")
        return True

    def number(self, value: Any, path: str) -> bool:
        if isinstance(value, bool) or not isinstance(value, (int, float)):
            return self.fail(path, "a finite number was required")
        if value != value or value in (float("inf"), float("-inf")):
            return self.fail(path, "a finite number was required")
        return True

    def whole(self, value: Any, path: str) -> bool:
        if not self.number(value, path):
            return False
        if float(value) != int(value):
            return self.fail(path, "a whole number was required")
        return True

    def string(self, value: Any, path: str) -> bool:
        if not isinstance(value, str):
            return self.fail(path, "a string was required")
        return True

    def boolean(self, value: Any, path: str) -> bool:
        if not isinstance(value, bool):
            return self.fail(path, "a true or false was required")
        return True

    def index(self, value: Any, path: str, size: int, table: str) -> bool:
        """A whole number that indexes a table, with the table's size in the message."""
        if not self.whole(value, path):
            return False
        if value < 0 or value >= size:
            return self.fail(path, f"{value} is outside {table}, which holds {size}")
        return True

    def one(self, value: Any, path: str, allowed: Sequence[str]) -> bool:
        if not self.string(value, path):
            return False
        if value not in allowed:
            return self.fail(path, f"{value} is not one of {', '.join(allowed)}")
        return True

    def nullable(self, value: Any, path: str, check: Callable[[Any, str], bool]) -> bool:
        """A field that may be a value or null, which is a value a script could write."""
        return True if value is None else check(value, path)


def check_colour(shape: ShapeCheck, value: Any, path: str) -> bool:
    if not shape.array(value, path):
        return False
    if len(value) != 4:
        return shape.fail(path, "a colour is four numbers")
    for at in range(3):
        if not shape.whole(value[at], f"{path}[{at}]"):
            return False
        channel = value[at]
        if channel < 0 or channel > 255:
            return shape.fail(f"{path}[{at}]", f"a channel runs 0 to 255 and this is {channel}")
    if not shape.number(value[3], f"{path}[3]"):
        return False
    alpha = value[3]
    if alpha < 0 or alpha > 1:
        return shape.fail(f"{path}[3]", f"an alpha runs 0 to 1 and this is {alpha}")
    return True


def check_constant(shape: ShapeCheck, value: Any, path: str) -> bool:
    """One constant pool entry, section 2.9: a two element array of tag and value."""
    if not shape.array(value, path):
        return False
    if len(value) != 2:
        return shape.fail(path, "a pool entry is a tag and a value")
    tag = value[0]
    if not shape.one(tag, f"{path}[0]", CONSTANT_TAGS):
        return False
    held = value[1]
    if tag == "z":
        return True if held is None else shape.fail(f"{path}[1]", "the absent entry holds null")
    if tag == "b":
        return shape.boolean(held, f"{path}[1]")
    if tag == "n":
        return shape.number(held, f"{path}[1]")
    if tag == "s":
        return shape.string(held, f"{path}[1]")
    return check_colour(shape, held, f"{path}[1]")


def check_field(shape: ShapeCheck, value: Any, path: str, keys: Set[str]) -> bool:
    """A declaration field, section 2.3: a value, or the input reference.

    The key half of check 10 lives here, because this is the one walk that visits
    every field that may hold a reference. The value half runs later, with the
    host's settings in hand, and is OS6019 rather than OS6018: that value came
    from a settings dialog and the user who typed it can correct it.
    """
    if value is None or isinstance(value, (bool, str)):
        return True
    if isinstance(value, (int, float)):
        return shape.number(value, path)
    if isinstance(value, list):
        for at, one in enumerate(value):
            if not shape.number(one, f"{path}[{at}]"):
                return False
        return True
    if not shape.object(value, path):
        return False
    if "input" not in value:
        return shape.fail(path, "an object here is an input reference")
    key = value["input"]
    if not shape.string(key, f"{path}.input"):
        return False
    if key not in keys:
        return shape.fail(path, f"it names the input {key}, which inputs[] does not declare")
    return True
